### DISCUSSION TITLE
testbench layout/design fo cosimulation of RTL and Simulink

### DISCUSSION THREAD

**Discussion by richardc87** [cocotb/cocotb#3311] [Q&A]
I would like to use cocotb to verify auto-generated HDL to Simulink model.  I have successfully created a python script that uses the Matlab API Engine for Python to load the Simulink model, simulate it, and capture data from the workspace.  I have also separately used cocotb to simulate the RTL using Questa.  I would now like to combine these two sims under one python script to compare outputs.  However, I am new to cocotb and would like some suggestions on how to layout the test-bench.  I'd like to hear how to take advantage of the cocotb package, the co-routine aspect of cocotb, and how to be efficient on my processor resources so that it can run Questa and Matlab engine.  My initial thought is to run the Matlab engine to collect data in one routine/test.  Then a second routine/test would run the RTL simulation.  Finally, the third routine/test would compare the outputs.  Thanks in advance.

**Comment by ktbarrett** [comment]
A more optimal approach would involve you running the Matlab model in parallel to the simulation using Python's `threading` library. This would all be done in a single test. At the end of the test use `thread.join()` to wait until the Matlab model function is finished (if it isn't already) then compare the results of the Matlab model against the simulation.
